google/resource_google_project_iam_member.go

```diff
@@ -0,0 +1,182 @@
+				ForceNew: true,
+			},
+			"member": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"etag": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceGoogleProjectIamMemberCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	pid, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	// Get the binding in the template
+	log.Println("[DEBUG]: Reading google_project_iam_member")
+	p := getResourceIamMember(d)
+	mutexKV.Lock(projectIamMemberMutexKey(pid, p.Role, p.Members[0]))
+	defer mutexKV.Unlock(projectIamMemberMutexKey(pid, p.Role, p.Members[0]))
+
+	err = projectIamPolicyReadModifyWrite(d, config, pid, func(ep *cloudresourcemanager.Policy) error {
+		// find the binding
+		var binding *cloudresourcemanager.Binding
+		for _, b := range ep.Bindings {
+			if b.Role != p.Role {
+				continue
+			}
+			binding = b
+			break
+		}
+		if binding == nil {
+			binding = &cloudresourcemanager.Binding{
+				Role:    p.Role,
+				Members: p.Members,
+			}
+		}
+
+		// Merge the bindings together
+		ep.Bindings = mergeBindings(append(ep.Bindings, p))
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	d.SetId(pid + ":" + p.Role + ":" + p.Members[0])
+	return resourceGoogleProjectIamMemberRead(d, meta)
+}
+
+func resourceGoogleProjectIamMemberRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	pid, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	eMember := getResourceIamMember(d)
+
+	log.Println("[DEBUG]: Retrieving policy for project", pid)
+	p, err := getProjectIamPolicy(pid, config)
+	if err != nil {
+		return err
+	}
+	log.Printf("[DEBUG]: Retrieved policy for project %q: %+v\n", pid, p)
+
+	var binding *cloudresourcemanager.Binding
+	for _, b := range p.Bindings {
+		if b.Role != eMember.Role {
+			continue
+		}
+		binding = b
+		break
+	}
+	if binding == nil {
+		d.SetId("")
+		return nil
+	}
+	var member string
+	for _, m := range binding.Members {
+		if m == eMember.Members[0] {
+			member = m
+		}
+	}
+	if member == "" {
+		d.SetId("")
+		return nil
+	}
+	d.Set("etag", p.Etag)
+	d.Set("member", member)
+	d.Set("role", binding.Role)
+	return nil
+}
+
+func resourceGoogleProjectIamMemberDelete(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	pid, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	member := getResourceIamMember(d)
+	mutexKV.Lock(projectIamMemberMutexKey(pid, member.Role, member.Members[0]))
+	defer mutexKV.Unlock(projectIamMemberMutexKey(pid, member.Role, member.Members[0]))
+
+	err = projectIamPolicyReadModifyWrite(d, config, pid, func(p *cloudresourcemanager.Policy) error {
+		bindingToRemove := -1
+		for pos, b := range p.Bindings {
+			if b.Role != member.Role {
+				continue
+			}
+			bindingToRemove = pos
+			break
+		}
+		if bindingToRemove < 0 {
+			return nil
```

Comment:
Because this happens in the read-modify-write helper function, we don't need to call SetId, because when it returns nil, it just returns back up to the Delete function. The Delete calls read, which will verify the member does not actually exist anymore, and then removes it from state.